packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -57,63 +58,137 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
-  // We set `autoUncompress` to false to ensure that we can trust the value of
-  // the `Content-Length` HTTP header. We automatically uncompress the content
-  // in our call to [consolidateHttpClientResponseBytes].
-  static final HttpClient _sharedHttpClient = HttpClient()..autoUncompress = false;
-
-  static HttpClient get _httpClient {
-    HttpClient client = _sharedHttpClient;
-    assert(() {
-      if (debugNetworkImageHttpClientProvider != null)
-        client = debugNetworkImageHttpClientProvider();
-      return true;
-    }());
-    return client;
-  }
+  // For [_pendingLoader] we don't need the value(worker isolate), just the future
+  // itself that is used as an indicator that successive load requests should be
+  // added to the list of pending load requests [_pendingLoadRequests].
+  static Future<void> _pendingLoader;
+  static RawReceivePort _loaderErrorHandler;
+  static List<_DownloadRequest> _pendingLoadRequests;
+  static SendPort _requestPort;
 
   Future<ui.Codec> _loadAsync(
     NetworkImage key,
     StreamController<ImageChunkEvent> chunkEvents,
   ) async {
+    RawReceivePort downloadResponseHandler;
     try {
       assert(key == this);
 
       final Uri resolved = Uri.base.resolve(key.url);
-      final HttpClientRequest request = await _httpClient.getUrl(resolved);
-      headers?.forEach((String name, String value) {
-        request.headers.add(name, value);
+
+      final Completer<TransferableTypedData> bytesCompleter = Completer<TransferableTypedData>();
+      downloadResponseHandler = RawReceivePort((_DownloadResponse response) {
+        if (response.bytes != null) {
+          if (bytesCompleter.isCompleted) {
+            // If an uncaught error occurred in the worker isolate, we'll have
+            // already completed our bytes completer.
+            return;
+          }
+          bytesCompleter.complete(response.bytes);
+        } else if (response.chunkEvent != null) {
+          chunkEvents.add(response.chunkEvent);
+        } else if (response.error != null) {
+          bytesCompleter.completeError(response.error);
+        } else {
+          assert(false);
+        }
       });
-      final HttpClientResponse response = await request.close();
-      if (response.statusCode != HttpStatus.ok)
-        throw image_provider.NetworkImageLoadException(statusCode: response.statusCode, uri: resolved);
-
-      final Uint8List bytes = await consolidateHttpClientResponseBytes(
-        response,
-        onBytesReceived: (int cumulative, int total) {
-          chunkEvents.add(ImageChunkEvent(
-            cumulativeBytesLoaded: cumulative,
-            expectedTotalBytes: total,
-          ));
-        },
+
+      // This will keep references to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release builds.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
```

Comment:
@goderbauer you were talking about how this PR breaks the ability to use [HttpOverrides](https://api.dartlang.org/stable/2.5.0/dart-io/HttpOverrides-class.html), but callers should be able to use `debugNetworkImageHttpClientProvider` to hook into an overridden HTTP client, no?